flowmanager/controllers.py

```diff
@@ -91,19 +91,30 @@ def upload(token, contents,
             permissions = verifyer.extract_permissions(token)
             if permissions and permissions.get('userid') == owner:
                 limits = permissions.get('permissions')
+                is_private = findability_getter(contents) == 'private'
                 max_datasets = limits.get('max_dataset_num', 0)
+                max_storage = limits.get(
+                    'max_private_storage_mb' if is_private else 'max_public_storage_mb', 0
+                )
                 current_datasets = registry.num_datasets_for_owner(owner)
+                fm = registry.file_manager()
+                current_storage = fm.get_total_size_for_owner(
+                    owner, 'private' if is_private else None
+                )
                 dataset_id = registry.format_identifier(owner, dataset_getter(contents))
                 is_revision = registry.get_dataset(dataset_id) is not None
-                if current_datasets < max_datasets or is_revision:
+                if not (current_datasets < max_datasets or is_revision):
+                    errors.append('Max datasets for user exceeded plan limit (%d)' % max_datasets)
+                elif max_storage < current_storage:
```

Comment:
test should be whether (current_storage + new file size < limit).
This way I can upload many GBs beyond the limit...